File: BAEKJOON/14502.cpp
/*
https://www.acmicpc.net/problem/14502
*/


#include <cstdio>
#include <iostream>
#include <queue>
#include <vector>
#include <memory.h>


using namespace std;

struct point{
    int x; int y;
};

int dx[4] = {0,0,1,-1};
int dy[4] = {1,-1,0,0};

int arr[8][8];
bool chk[8][8];
int n,m;
int safeZero = -3;//초기 안전영역 (벽 3개 세울거니까)
int result = 0;

vector<point> virus;
vector<point> safe;

int BFS(){
    //바이러스 퍼트리면서 안전영역 구하기
    //퍼진 개수 return

    int countVirus = 0;
    queue<point> q;

    for(int i=0; i<virus.size(); i++){
        int x = virus[i].x;
        int y = virus[i].y;

        chk[x][y] = true;
        point p = {x,y};
        q.push(p);

    }

    while(q.empty() == false){
        int cx = q.front().x;
        int cy = q.front().y; q.pop();

        for(int i=0; i<4; i++){
            int nx = cx+dx[i];
            int ny = cy+dy[i];

            if(nx>=0 && ny>=0 && nx<n && ny<m && arr[nx][ny] ==0 && chk[nx][ny] == false){
                countVirus++;
                chk[nx][ny] = true;
                point np = {nx,ny};
                q.push(np);
            }
        }
    }

    return countVirus;
}

void choice(int x, int k){
    //벽 3개 세우기
    if(x == 3){
        
        memset(chk,false,sizeof(chk));
        int safeTemp = safeZero;

        safeTemp-=BFS();
        

        if(result<safeTemp){
            result = safeTemp;
        }
        

    }else{
        for(int i=k; i<safe.size(); i++){
            arr[safe[i].x][safe[i].y] = 1;
            choice(x+1,i+1);
            arr[safe[i].x][safe[i].y] = 0;
        }
    }
}


int main(){
    cin>>n>>m;
    
    for(int i=0; i<n; i++){
        for(int j=0 ; j<m; j++){
            cin>>arr[i][j];
            if(arr[i][j] == 0){
                point temp = {i,j};
                safe.push_back(temp);
                safeZero++;

            }else if(arr[i][j] == 2){
                point temp = {i,j};
                virus.push_back(temp);
            }
        }
    }

    choice(0,0);

    cout<<result;

    return 0;
}

File: BAEKJOON/JAVA/14502.java
import java.util.*;
import java.io.*;
/*
https://www.acmicpc.net/problem/14502
첫째 줄에 지도의 세로 크기 N과 가로 크기 M이 주어진다. (3 ≤ N, M ≤ 8)
둘째 줄부터 N개의 줄에 지도의 모양이 주어진다. 0은 빈 칸, 1은 벽, 2는 바이러스가 있는 위치이다. 2의 개수는 2보다 크거나 같고, 10보다 작거나 같은 자연수이다.
빈 칸의 개수는 3개 이상이다.

1. 입력
2. 임의의 벽 3개 할당
3. 바이러스 확산 및 카운팅
*/
class Point {
    int x;
    int y;

    Point(int x, int y){
        this.x = x;
        this.y = y;
    }
}

public class Main {
    public static int n, m;
    public static String[][]arr = new String[10][10];
    public static Boolean[][]chk = new Boolean[10][10];
    public static int[] dx = {0,0,1,-1};
    public static int[] dy = {1,-1,0,0};
    public static int result = 0;
    public static int totSafe = 0;
    public static ArrayList<Point> v= new ArrayList<>();
    public static ArrayList<Point> virus = new ArrayList<>();
        
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        
        for(int i=0; i<n; i++){
            String[] str = br.readLine().split(" ");
            for(int j=0; j<m; j++){
                arr[i][j] = str[j];
                if("0".equals(arr[i][j])){
                    v.add(new Point(i, j));
                    totSafe++;
                }else if("2".equals(arr[i][j])){
                    virus.add(new Point(i, j));
                }
            }
        }
        
        //1. 임의의 벽 3개 할당
        totSafe = totSafe-3;
        DFS(0);

        System.out.println(result);

    }

    static public void DFS(int x) {
        if(x == 3){
            String[][]tmp = arr.clone();
            for(int i=0; i<n; i++){
                tmp[i] = arr[i].clone();
            }

            int tmpRes = spreadVirus();
            if(result < tmpRes) result = tmpRes;

            for(int i=0; i<n; i++){
                arr[i] = tmp[i].clone();
            }

        }else{
            for(int i=0; i<v.size(); i++){
                if("0".equals(arr[v.get(i).x][v.get(i).y])){
                    arr[v.get(i).x][v.get(i).y] = "1";
                    DFS(x+1);
                    arr[v.get(i).x][v.get(i).y] = "0";
                }
            }
        }
    }

    static public int spreadVirus() {
        int safe = totSafe;
        Queue<Point> q = new LinkedList<>();

        for(int i=0; i<virus.size(); i++){
            q.add(virus.get(i));
        }

        while(q.isEmpty() == false){
            Point p = q.poll();
            int cx = p.x;
            int cy = p.y;
            for(int i=0; i<4; i++){
                int nx = cx+dx[i];
                int ny = cy+dy[i];
                if( nx<0 || ny<0 || nx>=n || ny>=m || "0".equals(arr[nx][ny]) == false) continue;
                else{
                    arr[nx][ny] = "2";
                    q.add(new Point(nx,ny));
                    safe--;
                }
            }
        }
        return safe;
    }
}
